File: pwa/src/lib/protocol/aead.ts
import { chacha20poly1305 } from "@noble/ciphers/chacha.js";

export const MAX_PLAINTEXT = 262116;
export const MAX_PAYLOAD = 262144;
export const DIR_C = 0x63; // 'c'
export const DIR_S = 0x73; // 's'

export function aad(routeId: Uint8Array, dir: number): Uint8Array {
  if (!(routeId instanceof Uint8Array) || routeId.length !== 16) throw new Error("route_id must be 16 bytes");
  if (dir !== DIR_C && dir !== DIR_S) throw new Error("invalid AEAD direction");
  const out = new Uint8Array(21);
  out[0] = 0x01;
  out[1] = 0x05;
  out.set(routeId, 4);
  out[20] = dir;
  return out;
}

export class Direction {
  seq = 0n;
  constructor(
    public key: Uint8Array,
    public dir: number,
  ) {}

  seal(routeId: Uint8Array, plaintext: Uint8Array): Uint8Array {
    if (!(plaintext instanceof Uint8Array)) throw new Error("plaintext must be bytes");
    if (plaintext.length > MAX_PLAINTEXT) throw new Error("plaintext exceeds 262116");
    if (this.seq < 0n || this.seq > 0xffff_ffff_ffff_ffffn) throw new Error("aead seq overflow");
    const nonce = new Uint8Array(12);
    new DataView(nonce.buffer).setBigUint64(4, this.seq, false);
    const cipher = chacha20poly1305(this.key, nonce, aad(routeId, this.dir));
    const ct = cipher.encrypt(plaintext);
    const out = new Uint8Array(12 + ct.length);
    out.set(nonce, 0);
    out.set(ct, 12);
    this.seq++;
    return out;
  }

  open(routeId: Uint8Array, payload: Uint8Array): Uint8Array {
    if (!(payload instanceof Uint8Array) || payload.length < 28 || payload.length > MAX_PAYLOAD) {
      throw new Error("aead payload length");
    }
    const nonce = payload.slice(0, 12);
    const rest = payload.slice(12);
    if (nonce[0] !== 0 || nonce[1] !== 0 || nonce[2] !== 0 || nonce[3] !== 0) throw new Error("aead seq mismatch");
    const got = new DataView(nonce.buffer, nonce.byteOffset, 12).getBigUint64(4, false);
    if (got !== this.seq) throw new Error("aead seq mismatch");
    const cipher = chacha20poly1305(this.key, nonce, aad(routeId, this.dir));
    const pt = cipher.decrypt(rest);
    this.seq++;
    return pt;
  }
}

File: internal/crypto/aead/aead.go
// Package aead implements Pairfob FWD AEAD: ChaCha20-Poly1305, 12-byte nonce on wire, 21-byte AAD.
package aead

import (
	"encoding/binary"
	"errors"

	"golang.org/x/crypto/chacha20poly1305"
)

const (
	MaxPlaintext = 262116
	MaxPayload   = 262144 // nonce(12)+pt+tag(16)
	DirClient    = byte('c')
	DirServer    = byte('s')
)

var ErrTooLarge = errors.New("plaintext exceeds 262116")
var ErrSeq = errors.New("aead seq mismatch")
var ErrOpen = errors.New("aead open failed")

type Direction struct {
	Key []byte
	Seq uint64
	Dir byte
}

func Seal(d *Direction, routeID [16]byte, plaintext []byte) ([]byte, error) {
	if len(plaintext) > MaxPlaintext {
		return nil, ErrTooLarge
	}
	aead, err := chacha20poly1305.New(d.Key)
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, 12)
	binary.BigEndian.PutUint64(nonce[4:], d.Seq)
	aad := AAD(0x01, 0x05, 0, routeID, d.Dir)
	ct := aead.Seal(nil, nonce, plaintext, aad)
	out := append(nonce, ct...)
	d.Seq++
	return out, nil
}

func Open(d *Direction, routeID [16]byte, payload []byte) ([]byte, error) {
	if len(payload) < 12+16 {
		return nil, ErrOpen
	}
	nonce, rest := payload[:12], payload[12:]
	got := binary.BigEndian.Uint64(nonce[4:])
	if got != d.Seq {
		return nil, ErrSeq
	}
	if nonce[0]|nonce[1]|nonce[2]|nonce[3] != 0 {
		return nil, ErrSeq
	}
	aead, err := chacha20poly1305.New(d.Key)
	if err != nil {
		return nil, err
	}
	aad := AAD(0x01, 0x05, 0, routeID, d.Dir)
	pt, err := aead.Open(nil, nonce, rest, aad)
	if err != nil {
		return nil, ErrOpen
	}
	d.Seq++
	return pt, nil
}

func AAD(version, typ byte, flags uint16, routeID [16]byte, dir byte) []byte {
	aad := make([]byte, 21)
	aad[0] = version
	aad[1] = typ
	binary.BigEndian.PutUint16(aad[2:4], flags)
	copy(aad[4:20], routeID[:])
	aad[20] = dir
	return aad
}
